File: examples/baas-interaction-examples/reserve-items/src/main/scala/webshop/webservice/ReserveItems.scala
package webshop.webservice

import scala.concurrent.Future
import com.ing.baker.recipe.javadsl.Interaction

trait ReserveItems extends Interaction {

  case class OrderId(orderId: String)

  case class Item(itemId: String)

  case class OrderPlaced(orderId: OrderId, items: List[Item])

  case class ReservedItems(items: List[Item], data: Array[Byte])

  sealed trait ReserveItemsOutput

  case class OrderHadUnavailableItems(unavailableItems: List[Item]) extends ReserveItemsOutput

  case class ItemsReserved(reservedItems: ReservedItems) extends ReserveItemsOutput

  def apply(orderId: OrderId, items: List[Item]): Future[ReserveItemsOutput]
}

File: core/baker-test/src/test/java/com/ing/baker/test/javadsl/recipe/ReserveItems.java
package com.ing.baker.test.javadsl.recipe;

import com.ing.baker.recipe.annotations.AsyncInteraction;
import com.ing.baker.recipe.annotations.FiresEvent;
import com.ing.baker.recipe.annotations.RequiresIngredient;
import com.ing.baker.recipe.javadsl.Interaction;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface ReserveItems extends Interaction {
    @AsyncInteraction
    @FiresEvent(oneOf = {ItemsReserved.class, OrderHadUnavailableItems.class})
    CompletableFuture<ReserveItemsOutput> apply(
            @RequiresIngredient("orderId") String orderId,
            @RequiresIngredient("items") List<String> items);

    interface ReserveItemsOutput {}
    class ItemsReserved implements ReserveItemsOutput {}
    class OrderHadUnavailableItems implements ReserveItemsOutput {}
}
